src/librustc/infer/error_reporting/named_anon_conflict.rs

```diff
@@ -0,0 +1,115 @@
+    // &hir::Arg of the function argument corresponding to the anonymous
+    // region and the Ty corresponding to the named region.
+    // Currently only the case where the function declaration consists of
+    // one named region and one anonymous region is handled.
+    // Consider the example `fn foo<'a>(x: &'a i32, y: &i32) -> &'a i32`
+    // Here, the `y` and the `Ty` of `y` is returned after being substituted
+    // by that of the named region.
+    pub fn find_arg_with_anonymous_region(&self,
+                                          anon_region: Region<'tcx>,
+                                          named_region: Region<'tcx>)
+                                          -> Option<(&hir::Arg, ty::Ty<'tcx>)> {
+
+        match *anon_region {
+            ty::ReFree(ref free_region) => {
+
+                let id = free_region.scope;
+                let node_id = self.tcx.hir.as_local_node_id(id).unwrap();
+                let body_id = self.tcx.hir.maybe_body_owned_by(node_id).unwrap();
+
+                let body = self.tcx.hir.body(body_id);
+                body.arguments
+                    .iter()
+                    .filter_map(|arg| if let Some(tables) = self.in_progress_tables {
+                                    let ty = tables.borrow().node_id_to_type(arg.id);
+                                    let mut found_anon_region = false;
+                                    let new_arg_ty = self.tcx
+                                        .fold_regions(&ty,
+                                                      &mut false,
+                                                      |r, _| if *r == *anon_region {
+                                                          found_anon_region = true;
+                                                          named_region
+                                                      } else {
+                                                          r
+                                                      });
+                                    if found_anon_region {
+                                        return Some((arg, new_arg_ty));
+                                    } else {
+                                        None
+                                    }
+                                } else {
+                                    None
+                                })
+                    .next()
+            }
+            _ => None,
+        }
+
+    }
+
+    // This method generates the error message for the case when
+    // the function arguments consist of a named region and an anonymous
+    // region and corresponds to `ConcreteFailure(..)`
+    pub fn report_named_anon_conflict(&self, error: &RegionResolutionError<'tcx>) -> bool {
+
+        let (span, sub, sup) = match *error {
+            ConcreteFailure(ref origin, sub, sup) => (origin.span(), sub, sup),
+            _ => return false, // inapplicable
+        };
+
+        let (named, (var, new_ty)) =
```

Comment:
let's rename `var` to `arg`, for consistency. Probably a good idea to put a comment before this block of code, too:

```
// Determine whether the `sub` and `sup` consist of one named region (`'a`)
// and one anonymous (elided) region. If so, find the parameter `arg` where the
// anonymous region appears (there must always be one; we only introduced
// anonymous regions in parameters) as well as a version `new_ty` of its type
// where the anonymous region is replaced with the named one.
```